src/DocumentDB.ChangeFeedProcessor.IntegrationTests/V3CompatibilityTests.cs

```diff
@@ -0,0 +1,273 @@
+                    .WithObserverFactory(observerFactory)
+                    .WithHostName("smoke_test")
+                    .WithFeedCollection(this.MonitoredCollectionInfo)
+                    .WithLeaseCollection(this.LeaseCollectionInfo)
+                    .BuildAsync();
+
+            await changeFeedProcessorBuilder.StartAsync();
+            await this.WaitUntilLeaseStoreIsInitializedAsync();
+            await changeFeedProcessorBuilder.StopAsync();
+
+            // Verify that no leases have LeaseToken (V3 contract)
+            int leasesProcessed = 0;
+            using (DocumentClient client = new DocumentClient(this.LeaseCollectionInfo.Uri, this.LeaseCollectionInfo.MasterKey, this.LeaseCollectionInfo.ConnectionPolicy))
+            {
+                Uri collectionUri = UriFactory.CreateDocumentCollectionUri(this.LeaseCollectionInfo.DatabaseName, this.LeaseCollectionInfo.CollectionName);
+
+                IDocumentQuery<JObject> query = client.CreateDocumentQuery<JObject>(collectionUri, "SELECT * FROM c").AsDocumentQuery();
+                while (query.HasMoreResults)
+                {
+                    foreach (JObject lease in await query.ExecuteNextAsync())
+                    {
+                        string leaseId = lease.Value<string>("id");
+                        if (leaseId.Contains(".info"))
+                        {
+                            // These are the store initialization marks
+                            continue;
+                        }
+
+                        Assert.NotNull(lease.Value<string>("PartitionId"));
+                        Assert.Null(lease.Value<string>("LeaseToken"));
+                        leasesProcessed++;
+                    }
+                }
+            }
+
+            Assert.True(leasesProcessed > 0);
+        }
+
+        /// <summary>
+        /// When the user migrates from V3 CFP, the leases contain LeaseToken.
+        /// To allow for backward compatibility (V2 -> V3 -> V2) we need to honor the existence of LeaseToken and maintain its value in case the lease was created by V3.
+        /// </summary>
+        [Fact]
+        public async Task Schema_OnV2MigrationMaintainLeaseToken()
+        {
+            List<int> expectedIds = Enumerable.Range(0, 20).ToList();
+            ManualResetEvent firstSetOfResultsProcessed = new ManualResetEvent(false);
+            ManualResetEvent secondSetOfResultsProcessed = new ManualResetEvent(false);
+            List<int> receivedIds = new List<int>();
+            TestObserverFactory observerFactory = new TestObserverFactory(
+                (FeedProcessing.IChangeFeedObserverContext context, IReadOnlyList<Document> docs) =>
+                {
+                    foreach (Document doc in docs)
+                    {
+                        receivedIds.Add(int.Parse(doc.Id));
+                    }
+
+                    if (receivedIds.Count == 10)
```

Comment:
ideally, 10 should be a constant. 20 would be 10*2.

Reply:
Done